### PR TITLE
feat: enable optional fields

### LINKED ISSUE #1389: Optional text fields are still required
### Issue Description

Even though text fields are not set to required, for filling out a document with a text field still requires to click and fill each field to continue. This even happens when adding a pre-filled value to a field.

### Steps to Reproduce

Add a text field to any document

### Expected Behavior

Text field not marked as required will not require input

### Current Behavior

Text field not marked as required does require input

### Screenshots (optional)

_No response_

### Operating System [e.g., Windows 10]

_No response_

### Browser [e.g., Chrome, Firefox]

_No response_

### Version [e.g., 2.0.1]

1.7.1-rc.3

### Please check the boxes that apply to this issue report.

- [X] I have searched the existing issues to make sure this is not a duplicate.
- [X] I have provided steps to reproduce the issue.
- [ ] I have included relevant environment information.
- [ ] I have included any relevant screenshots.
- [ ] I understand that this is a voluntary contribution and that there is no guarantee of resolution.
- [ ] I want to work on creating a PR for this issue if approved

### PR DESCRIPTION
**Pull request by catalinpit**
fixes #1389

## Release Notes

- **New Features**
	- Enhanced validation logic for form fields in the signing process, ensuring only relevant fields are checked.
	- Improved completion status determination for signing fields, accommodating optional advanced fields.
	- Added advanced field type checks to streamline document sealing and signing processes.

- **Bug Fixes**
	- Refined error messages for unsigned fields to clarify issues related to recipients.

- **Documentation**
	- Updated comments for better understanding of validation logic and field handling.

### CODE REVIEW

File: packages/lib/jobs/definitions/internal/seal-document.ts [hunk 1]
```diff
@@ -141,7 +152,9 @@ export const SEAL_DOCUMENT_JOB_DEFINITION = {
       }
 
       for (const field of fields) {
-        await insertFieldInPDF(pdfDoc, field);
+        if (field.inserted) {
```
Comment (catalinpit) [inline packages/lib/jobs/definitions/internal/seal-document.ts:155]: question: should we remove the uninserted (optional) fields from the documents altogether? at the moment, they're not inserted in the PDF, but they are still existing in the db and are linked to the document.

File: apps/web/src/app/(signing)/sign/[token]/form.tsx [hunk 2]
```diff
@@ -56,19 +58,29 @@ export const SigningForm = ({
   // Keep the loading state going if successful since the redirect may take some time.
   const isSubmitting = formState.isSubmitting || formState.isSubmitSuccessful;
 
+  const fieldsRequiringValidation = useMemo(
+    () =>
+      fields.filter((field) => {
+        const isRequired = field.fieldMeta && ZFieldMetaSchema.parse(field.fieldMeta)?.required;
+
+        return isRequired || !isAdvancedField(field.type) ? !field.inserted : false;
```
Comment (Mythie) [inline apps/web/src/app/(signing)/sign/[token]/form.tsx:66]: can we simplify this statement. I'd rather have it over multiple lines than the current `||` into a ternary 😄

File: apps/web/src/app/(signing)/sign/[token]/sign-dialog.tsx [hunk 3]
```diff
@@ -37,7 +39,14 @@ export const SignDialog = ({
...
   const [showDialog, setShowDialog] = useState(false);
   const truncatedTitle = truncateTitle(documentTitle);
-  const isComplete = fields.every((field) => field.inserted);
+
+  const isComplete = fields.every((field) => {
+    const isOptionalAdvancedField =
+      isAdvancedField(field.type) &&
+      (!field.fieldMeta || !ZFieldMetaSchema.parse(field.fieldMeta)?.required);
+
+    return isOptionalAdvancedField || field.inserted;
```
Comment (Mythie) [inline apps/web/src/app/(signing)/sign/[token]/sign-dialog.tsx:48]: this is probably better as 

```
if (isOptionalAdvancedField) {
  return true
}

return field.inserted
```

File: packages/lib/jobs/definitions/internal/seal-document.ts (hunk 1 above)
Comment (Mythie) [reply packages/lib/jobs/definitions/internal/seal-document.ts:155]: they should remain

**Review by Mythie — CHANGES_REQUESTED** [review]
a lot of ternaries in here that use negation, I'd avoid this out of habit just because it can be a bit more confusing in terms of branching when reading the code.

I'm fine with more verbose code that uses if statements and early returns instead :)

File: apps/web/src/app/(signing)/sign/[token]/sign-dialog.tsx [hunk 4]
```diff
@@ -37,7 +39,18 @@ export const SignDialog = ({
 }: SignDialogProps) => {
   const [showDialog, setShowDialog] = useState(false);
   const truncatedTitle = truncateTitle(documentTitle);
-  const isComplete = fields.every((field) => field.inserted);
+
+  const isComplete = fields.every((field) => {
+    const isOptionalAdvancedField =
+      isAdvancedField(field.type) &&
+      (!field.fieldMeta || !ZFieldMetaSchema.parse(field.fieldMeta)?.required);
```
[the remark this answers is not in the document]
Comment (dguyen) [reply apps/web/src/app/(signing)/sign/[token]/sign-dialog.tsx:46]: Maybe create a util that parses it for you or returns null, so you can reuse it in all the other functions you have here

Or even just create an `isOptionalField` util function

**Review by dguyen — APPROVED** [review]

**Pull request merged**